metrics/prometheus/metrics.go

```diff
@@ -1,42 +1,120 @@
-//				Help:        "Description",
-//				ConstLabels: map[string]string{
-//					"version":   version,
-//					"hash":      hash,
-//					"buildTime": buildTime,
-//				},
-//			},
-//			[]string{"endpoint"},
-//		),
-//	}
-//	prometheus.Register(pm.Counter)
-//  prometheus.Register(pm.ResponseTime)
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Metrics contains the metrics that will be recorded.
+// Rather than create metrics ad-hoc, we prefer having a central list of metrics,
+// as per the Prometheus "best practices" in their documentation.
+// See: https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics
+//
+// The labels associated with metrics should be exclusively descriptive.
+// For example, one might think that this is an applicable demonstration of "hydra_consent_requests_accepted":
+//   hydra_consent_requests_accepted{scope="scope:1"} 1
+//   hydra_consent_requests_accepted{scope="scope:2"} 1
+// when a consent request requests []string{"scope:1", "scope:2"}.
+// This is not good; if we were to query Prometheus for `sum(hydra_consent_requests_accepted)`, we would get 2, when in reality,
+// only 1 consent request was accepted. Instead, separate data like "scopes" into a more descriptive time series:
+//   hydra_consent_requests_accepted{scopes="2"} 1
+//   hydra_consent_requests_accepted_scopes{scope="scope:1"} 1
+//   hydra_consent_requests_accepted_scopes{scope="scope:2"} 1
+//
+// Some of these values will be populated by HTTP middleware, while others
+// will be populated elsewhere throughout the project
+type Metrics struct {
+	HTTPRequestCount      *prometheus.CounterVec
+	HTTPRequestDuration   *prometheus.HistogramVec
+	HTTPResponseSizeBytes *prometheus.HistogramVec
+
+	ConsentRequests         *prometheus.CounterVec
+	ConsentRequestScopes    *prometheus.CounterVec
+	ConsentRequestAudiences *prometheus.CounterVec
+	ConsentRequestsRejected *prometheus.CounterVec
+
+	LoginRequests        *prometheus.CounterVec
+	LoginRequestSubjects *prometheus.CounterVec
+
+	AccessTokensIssued   *prometheus.CounterVec
+	AccessTokensRevoked  *prometheus.CounterVec
+	RefreshTokensIssued  *prometheus.CounterVec
+	RefreshTokensRevoked *prometheus.CounterVec
+	IDTokensIssued       *prometheus.CounterVec
+
+	Grants  *prometheus.GaugeVec
+	Clients *prometheus.GaugeVec
+	Keys    *prometheus.GaugeVec
+}
+
+// NewMetrics registers our metrics and their labels to the prometheus Registry.
 func NewMetrics(version, hash, date string) *Metrics {
-	pm := &Metrics{}
+	pm := &Metrics{
+		HTTPRequestCount: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "http_request_count",
+			Help: "",
+		}, []string{"code", "method"}),
+		HTTPRequestDuration: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    "http_request_duration_seconds",
+				Help:    "A histogram of latencies for requests",
+				Buckets: []float64{.05, .25, .5, 1, 2.5, 5, 10},
```

Comment:
Is it possible to create auto buckets here? So that prometheus chooses the right bucket size and median itself?

Reply:
No unfortunately. It actually can become a problem with visualization if you have a sample that changes around its bucket sizes over time, so it's important that we at least have these values set the way we want them the first time around.

Reply:
Also worth noting, the lower bucket count, the better. Querying histograms, especially over a large time range can be pretty taxing on the Prometheus server. We run into this issue at work now where a histogram will time out if queried over a 30 day span because there are too many buckets.

The histogram will include `count` and `sum`, so no data will really be missing.

Reply:
Unless you are referring to this: :sweat_smile: 

https://godoc.org/github.com/prometheus/client_golang/prometheus#LinearBuckets

https://godoc.org/github.com/prometheus/client_golang/prometheus#ExponentialBuckets

Reply:
Hm yeah I think neither linear nor exponential make sense here.